File: OneDrive/Desktop/Third Perspective/GITSYNC/apps/api/src/jobs/autoInvest.ts
import cron from 'node-cron';
import { query } from '../config/database';
import { initiateSTKPush } from '../services/mpesa.service';

/**
 * Auto-invest job - runs daily at 01:00 EAT
 * Automatically invests a percentage of profits for opted-in users
 */
export function startAutoInvestJob(): void {
  cron.schedule('0 22 * * *', async () => {
    console.log('🤖 Running auto-invest...');

    try {
      // Find users with auto-invest enabled
      const users = await query(
        `SELECT u.id, u.phone, u.auto_invest_percentage, u.auto_invest_fund_id, f.name as fund_name
         FROM users u
         LEFT JOIN mmf_funds f ON u.auto_invest_fund_id = f.id
         WHERE u.auto_invest_enabled = true AND u.auto_invest_percentage > 0 AND u.auto_invest_fund_id IS NOT NULL`
      );

      for (const user of users.rows) {
        // Calculate today's profit
        const profitResult = await query(
          `SELECT COALESCE(SUM(amount), 0) as today_profit
           FROM transactions
           WHERE user_id = $1 AND type IN ('c2b', 'stk_push') AND status = 'completed'
             AND created_at >= CURRENT_DATE`,
          [user.id]
        );

        const todayProfit = parseFloat(profitResult.rows[0].today_profit);
        if (todayProfit <= 0) continue;

        const investAmount = Math.floor(todayProfit * (parseFloat(user.auto_invest_percentage) / 100));
        if (investAmount < 1000) continue; // Minimum investment threshold

        try {
          // Initiate STK Push for auto-investment
          const stkResponse = await initiateSTKPush({
            phone: user.phone,
            amount: investAmount,
            accountReference: `AUTO-${user.fund_name?.substring(0, 8) || 'INV'}`,
            description: `Auto-invest: ${user.auto_invest_percentage}% of daily profit`,
          });

          // Record transaction
          const txResult = await query(
            `INSERT INTO transactions (user_id, type, amount, phone, merchant_request_id, checkout_request_id, status, description)
             VALUES ($1, 'stk_push', $2, $3, $4, $5, 'pending', $6)
             RETURNING id`,
            [user.id, investAmount, user.phone, stkResponse.MerchantRequestID, stkResponse.CheckoutRequestID,
             `Auto-invest in ${user.fund_name}`]
          );

          // Create pending investment
          await query(
            `INSERT INTO investments (user_id, fund_id, amount, current_value, status, transaction_id)
             VALUES ($1, $2, $3, $3, 'pending', $4)`,
            [user.id, user.auto_invest_fund_id, investAmount, txResult.rows[0].id]
          );

          console.log(`  💰 Auto-invested KES ${investAmount} for user ${user.id}`);
        } catch (err) {
          console.error(`  ❌ Auto-invest failed for user ${user.id}:`, err);
        }
      }

      console.log(`✅ Auto-invest completed for ${users.rows.length} users`);
    } catch (error) {
      console.error('❌ Auto-invest job error:', error);
    }
  });

  console.log('⏰ Auto-invest job scheduled (daily at 01:00 EAT)');
}

File: OneDrive/Desktop/SharedProj/HazinaHub/backend/app/jobs/auto_invest.py
from datetime import datetime, date, timedelta
from apscheduler.schedulers.asyncio import AsyncIOScheduler

from app.models.user import User
from app.models.cashflow import Cashflow
from app.models.investment import Investment
from app.models.mmf_fund import MmfFund


async def run_auto_invest():
    """Auto-invest job — runs daily at 01:00 EAT (22:00 UTC)."""
    print("[INFO] Running auto-invest...")
    try:
        users = await User.find(
            User.auto_invest_enabled == True,
            User.auto_invest_percentage > 0,
        ).to_list()

        processed = 0
        for user in users:
            if not user.auto_invest_fund_id:
                continue

            # Today's profit (incoming cashflows since midnight UTC)
            today_start = datetime.utcnow().replace(hour=0, minute=0, second=0, microsecond=0)
            pipeline = [
                {"$match": {
                    "user_id": user.id,
                    "type": {"$in": ["inflow", "deposit", "return", "c2b", "stk_push"]},
                    "status": "completed",
                    "created_at": {"$gte": today_start},
                }},
                {"$group": {"_id": None, "total": {"$sum": "$amount"}}},
            ]
            result = await Cashflow.aggregate(pipeline).to_list()
            today_profit = result[0]["total"] if result else 0.0

            if today_profit <= 0:
                continue

            invest_amount = int(today_profit * (user.auto_invest_percentage / 100))
            if invest_amount < 1000:
                continue

            try:
                from beanie import PydanticObjectId
                fund = await MmfFund.get(PydanticObjectId(user.auto_invest_fund_id))
                if not fund:
                    continue

                tx = Cashflow(
                    user_id=user.id,
                    type="investment",
                    amount=invest_amount,
                    phone=user.phone,
                    status="completed",
                    description=f"Auto-invest tracking in {fund.name}",
                )
                await tx.insert()

                matures_at = None
                if fund.maturity_days > 0:
                    matures_at = datetime.utcnow() + timedelta(days=fund.maturity_days)

                inv = Investment(
                    user_id=user.id,
                    fund_id=fund.id,
                    amount=invest_amount,
                    current_value=invest_amount,
                    status="active",
                    invested_at=datetime.utcnow(),
                    matures_at=matures_at,
                    cashflow_id=tx.id,
                )
                await inv.insert()

                print(f"  [OK] Auto-invest tracked KES {invest_amount:,} for user {user.id} into {fund.name}.")
                processed += 1
            except Exception as e:
                print(f"  [ERROR] Auto-invest failed for user {user.id}: {e}")
        print(f"[OK] Auto-invest completed for {processed} users")
    except Exception as e:
        print(f"[ERROR] Auto-invest job error: {e}")


def register_auto_invest(scheduler: AsyncIOScheduler):
    # 22:00 UTC = 01:00 EAT
    scheduler.add_job(run_auto_invest, "cron", hour=22, minute=0, id="auto_invest")
    print("[SCHEDULED] Auto-invest job scheduled (daily at 01:00 EAT)")

